blog post about the K8SAudit EKS logs plugin (#721)
falcosecurity/falco-website: content/en/blog/falco-plugin-k8saudit-eks.md

---
title: "Monitoring your EKS clusters audit logs"
linktitle: "Monitoring your EKS clusters audit logs"
date: 2022-10-31
author: Thomas Labarussias
slug: k8saudit-eks-plugin
---

> At the beginning of this year, we introduced a game changing feature, [the Plugins](https://falco.org/blog/falco-announcing-plugins/). They allow Falco to monitor and trigger alert for any kind of events. We have been pleased to create plugins for [Github](https://github.com/falcosecurity/plugins/tree/master/plugins/github), [AWS Cloudtrail](https://github.com/falcosecurity/plugins/tree/master/plugins/cloudtrail), [Okta](https://github.com/falcosecurity/plugins/tree/master/plugins/okta) and replaced the way to consume the Audit Logs generated by a K8S Control Plane by a [dedicated plugin](https://github.com/falcosecurity/plugins/tree/master/plugins/k8saudit). We these plugin, Falco covers more and more aspect of you infrastructure and allows you to use a single syntax for rules.

Reviewer: ```suggestion
At the beginning of this year, Falco introduced a game changing feature, [Falco Plugins](https://falco.org/blog/falco-announcing-plugins/). They allow Falco to monitor and trigger alerts for any kind of event. We have since created plugins for [Github](https://github.com/falcosecurity/plugins/tree/master/plugins/github), [AWS CloudTrail](https://github.com/falcosecurity/plugins/tree/master/plugins/cloudtrail), [Okta](https://github.com/falcosecurity/plugins/tree/master/plugins/okta), and have replaced the way Falco consumes Audit Logs generated by a K8s API server through a [dedicated plugin](https://github.com/falcosecurity/plugins/tree/master/plugins/k8saudit). With these plugins, Falco covers more and more aspects of your infrastructure and allows you to use a single syntax for rules.
```

> At the beginning of this year, Falco introduced a game changing feature, [Falco Plugins](https://falco.org/blog/falco-announcing-plugins/). They allow Falco to monitor and trigger alerts for any kind of event. We have since created plugins for [Github](https://github.com/falcosecurity/plugins/tree/master/plugins/github), [AWS CloudTrail](https://github.com/falcosecurity/plugins/tree/master/plugins/cloudtrail), [Okta](https://github.com/falcosecurity/plugins/tree/master/plugins/okta), and have replaced the way Falco consumes Audit Logs generated by a K8s API server through a [dedicated plugin](https://github.com/falcosecurity/plugins/tree/master/plugins/k8saudit). With these plugins, Falco covers more and more aspects of your infrastructure and allows you to use a single syntax for rules.

Reviewer 2: At the beginning of this year, Falco introduced a game changing feature, [Falco Plugins](https://falco.org/blog/falco-announcing-plugins/). This new feature allows Falco to monitor and trigger alerts for any kind of event. Since the launch of the new plugin framework the Falco community has collaborated with  [Github](https://github.com/falcosecurity/plugins/tree/master/plugins/github), [AWS CloudTrail](https://github.com/falcosecurity/plugins/tree/master/plugins/cloudtrail), [Okta](https://github.com/falcosecurity/plugins/tree/master/plugins/okta) and created new plugins. This has replaced the way Falco consumes Audit Logs generated by a K8s API server through their [dedicated plugin](https://github.com/falcosecurity/plugins/tree/master/plugins/k8saudit). With these plugins, Falco covers more in depth the aspects of your infrastructure and allows you to use a single syntax for rules.

Author: ```suggestion
At the beginning of this year, Falco introduced a game changing feature, [Falco Plugins](https://falco.org/blog/falco-announcing-plugins/). This new feature allows Falco to monitor and trigger alerts for any kind of event. Since the launch of the new plugin framework the Falco community has collaborated with  [Github](https://github.com/falcosecurity/plugins/tree/master/plugins/github), [AWS CloudTrail](https://github.com/falcosecurity/plugins/tree/master/plugins/cloudtrail), and [Okta](https://github.com/falcosecurity/plugins/tree/master/plugins/okta) to create new plugins. This has replaced the way Falco consumes Audit Logs generated by a K8s API server through their [dedicated plugin](https://github.com/falcosecurity/plugins/tree/master/plugins/k8saudit). With these plugins, Falco covers more in depth the aspects of your infrastructure and allows you to use a single syntax for rules.
```

> For months (years...), our adopters asked us a way to monitor the [Audit logs of K8S](https://kubernetes.io/docs/tasks/debug/debug-cluster/audit/). The previous implementation used an internal webserver to receive the logs from the Control Plane. Even if it works well for manually installed and managed clusters, this method doesn't fit with the clusters managed by the Cloud Proviers, like EKS, AKS or GKE. For their own usages they capture the audit logs and expose them in their own log aggregator.

Reviewer: ```suggestion
For months (years...), our adopters have asked us for a way to monitor the [Audit Logs of K8s](https://kubernetes.io/docs/tasks/debug/debug-cluster/audit/). The previous implementation used an internal web server to receive the logs from the Kubernetes API. Even if it worked well for manually installed and managed clusters, this method didn't fit the clusters managed by the cloud providers, like EKS, AKS, or GKE. They capture the audit logs for their own usages and expose them in their own log aggregator.
```

> This situation is now solved thanks to the plugin framework and we're proud to announce the first release of the plugin for the EKS Audit logs !!!

Reviewer 2: This situation is now solved thanks to the plugin framework and we're proud to announce the first release of the plugin for EKS Audit logs!!!

## How it works

> AWS captures the Audit logs and expose them in their [Cloudwatch Logs](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/WhatIsCloudWatchLogs.html) service, we then have created libs to create a clean session to the AWS' API and pull logs from the relevant Cloudwatch Logs Stream. You can reuse these [libs](https://github.com/falcosecurity/plugins/tree/master/shared/go/aws) for any plugin you like to create for any Amazon's service.

Reviewer: ```suggestion
AWS captures the Audit Logs and exposes them in their [CloudWatch Logs](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/WhatIsCloudWatchLogs.html) service. We have made available libs to create a clean session to the AWS' API and pull logs from the relevant Cloudwatch Logs Stream. You can reuse these [libs](https://github.com/falcosecurity/plugins/tree/master/shared/go/aws) for any plugin you'd like to create for any Amazon service.
```

> AWS captures the Audit Logs and exposes them in their [CloudWatch Logs](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/WhatIsCloudWatchLogs.html) service. We have made available libs to create a clean session to the AWS' API and pull logs from the relevant Cloudwatch Logs Stream. You can reuse these [libs](https://github.com/falcosecurity/plugins/tree/master/shared/go/aws) for any plugin you'd like to create for any Amazon service.

Reviewer 2: AWS captures the audit logs and exposes them in the [CloudWatch Logs](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/WhatIsCloudWatchLogs.html) service. We have made available libs to create a clean session with the AWS API and pull logs from the relevant Cloudwatch Logs Stream. You can reuse these [libs](https://github.com/falcosecurity/plugins/tree/master/shared/go/aws) for any plugin you'd like to create for any Amazon service.

## Usage

The configuration for the usage of the plugin is:

```yaml
plugins:
  - name: k8saudit-eks
    library_path: libk8saudit-eks.so
    init_config:
      region: "us-east-1"
      profile: "default"
      shift: 10
      polling_interval: 10
      use_async: false
      buffer_size: 500
    open_params: "my-cluster"
  - name: json
    library_path: libjson.so
    init_config: ""

load_plugins: [k8saudit-eks, json]
```
With:
* `init_config`:
  * **profile**: The Profile to use to create the session, env var AWS_PROFILE if present
  * **region**: The Region of your EKS cluster, env var AWS_REGION is used if present
  * **use_async**: If true then async extraction optimization is enabled (Default: true)
  * **polling_interval**: Polling Interval in seconds (default: 5s)
  * **shift**: Time shift in past in seconds (default: 1s)
  * **buffer_size**: Buffer Size (default: 200)
* `open_params`: A string which contains the name of your EKS Cluster (required).

> If you run Falco inside an EKS cluster with a set up an OIDC provider, the **profile** and **region** can be omitted in favor of a `service account` + IAM Role (see [the official docs](https://docs.aws.amazon.com/eks/latest/userguide/enable-iam-roles-for-service-accounts.html))

Reviewer: ```suggestion
If you run Falco inside an EKS cluster with a setup of an OIDC provider, the **profile** and **region** parameters can be omitted in favor of a `service account` + IAM Role (see [the official docs](https://docs.aws.amazon.com/eks/latest/userguide/enable-iam-roles-for-service-accounts.html)).
```

## Default Rules

> A good thing with Kubernetes is it brings some standards in our industry. Apart few differences, the cluster work in the same way and produce the same formats of logs. This helps us a lot, by creating the [k8saudit plugin](https://github.com/falcosecurity/plugins/tree/master/plugins/k8saudit) we declared fields to extract and default rules, and we can reuse them for any plugin which consumes audit logs from a control plane. It a time saver for both the developers and adopters.

Reviewer: ```suggestion
A good thing about Kubernetes is that it brings some standards into our industry. Despite a few differences, the cluster works in the same way and produces the same formats of logs. This helps us enormously. By creating the [k8saudit plugin](https://github.com/falcosecurity/plugins/tree/master/plugins/k8saudit) we declared fields to extract, as well as some default rules, which we can reuse for any plugin that consumes Audit Logs from the API. It is a time saver for both, the developers and adopters.
```

> You can find the proposed default rules [here](https://github.com/falcosecurity/plugins/blob/master/plugins/k8saudit/rules/k8s_audit_rules.yaml). To use them just add and `alternative` required plugin at the beginning of the yaml file of rules like this:

Reviewer: ```suggestion
You can find the proposed default rules [here](https://github.com/falcosecurity/plugins/blob/master/plugins/k8saudit/rules/k8s_audit_rules.yaml). To use them just add an `alternative` required plugin at the beginning of the YAML file of rules like this:
```

```yaml
- required_engine_version: 15
- required_plugin_versions:
  - name: k8saudit
    version: 0.1.0
    alternatives:
      - name: k8saudit-eks
        version: 0.1.0
```

## Example

Let's try with a dummy rule, just to check it works:

```yaml
- required_engine_version: 15
- required_plugin_versions:
  - name: k8saudit-eks
    version: 0.1.0

- rule: Dummy rule
  desc: >
    Dummy rule
  condition: >
    ka.verb in (get,create,delete,update)
  output: user=%ka.user.name verb=%ka.verb target=%ka.target.name target.namespace=%ka.target.namespace resource=%ka.target.resource
  priority: WARNING
  source: k8s_audit
  tags: [k8s]
```

```shell
falco -c falco.yaml -r rules/dummy_k8s_audit_rules.yaml 
```

```shell
16:07:42.045023000: Warning user=eks:certificate-controller verb=get target=eks-certificates-controller target.namespace=kube-system resource=configmapsEvents detected: 1
Rule counts by severity:
   WARNING: 1
Triggered rules by rule name:
   Dummy rule: 1
Syscall event drop monitoring:
   - event drop detected: 0 occurrences
   - num times actions taken: 0
```

The referencies to `eks` proves it works!!!

## Disclaimer

> Our different tests noticed some latencies between the presence of the logs in the Cloudwatch Logs Stream and their evaluation by Falco, this is getting worse with highly requested Control Planes, the solution is to adapt the size of your nodes which run Falco, a minimal size of `2xlarge` seems a relevant security.

Reviewer: ```suggestion
Our tests noticed some latencies between the presence of the logs in the CloudWatch Logs Stream and their evaluation by Falco. This is more visible with highly requested API servers. The solution is to adapt the size of your nodes where Falco runs, considering a minimal size of `2xlarge` as a safe option.
```

> ### Conclusion

Reviewer: ```suggestion
## Conclusion
```

The official Helm chart will be updated as soon the `.so` will be available in the Falco images, stay tuned.

> With this first Plugin for a managed K8S solution, we hope to open the door for new contributions from the community for other flavors like GKE and AKS. If you need to create a plugin for another AWS services, you can also take a look at the libs we created to help the evelopments.

Reviewer: ```suggestion
With this first Plugin for a managed K8s solution, we hope to open the door for new contributions from the community for other flavors like GKE and AKS. If you need to create a plugin for another AWS service, take also a look at the libs we created to help the developers.
```
